components/post_view/channel_intro_message/messages/default.tsx

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+
+import {FormattedMessage} from 'react-intl';
+
+import AddGroupsToTeamModal from 'components/add_groups_to_team_modal';
+
+import {Permissions} from 'mattermost-redux/constants';
+
+import TeamPermissionGate from 'components/permissions_gates/team_permission_gate';
+import {t} from 'utils/i18n';
+import {Constants, ModalIdentifiers} from 'utils/constants';
+import ToggleModalButton from 'components/toggle_modal_button';
+
+import LocalizedIcon from 'components/localized_icon';
+import {Channel} from '@mattermost/types/channels';
+import ChannelPermissionGate from 'components/permissions_gates/channel_permission_gate';
+import {PluginComponent} from 'types/store/plugins';
+import FormattedMarkdownMessage from 'components/formatted_markdown_message';
+import AddMembersButton from '../add_members_button';
+import BoardsButton from '../boards_button';
+import SetHeaderButton from '../set_header_button';
+
+type Props = {
+    boardComponent?: PluginComponent;
+    channel: Channel;
+    enableUserCreation?: boolean;
+    isReadOnly?: boolean;
+    stats: any;
+    teamIsGroupConstrained?: boolean;
+    usersLimit: number;
+}
+
+const DefaultIntroMessage = ({
+    boardComponent,
+    channel,
+    enableUserCreation,
+    isReadOnly,
+    stats,
+    teamIsGroupConstrained,
+    usersLimit,
+}: Props) => {
+    let teamInviteLink = null;
+    const totalUsers = stats.total_users_count;
+    const isPrivate = channel.type === Constants.PRIVATE_CHANNEL;
+
+    let setHeaderButton = null;
+    let boardCreateButton = null;
+    if (!isReadOnly) {
+        boardCreateButton = (
+            <BoardsButton
+                channel={channel}
+                boardComponent={boardComponent}
+            />
+        );
+        const children = (
+            <SetHeaderButton
+                channel={channel}
+            />
+        );
+        if (children) {
+            setHeaderButton = (
+                <ChannelPermissionGate
+                    teamId={channel.team_id}
+                    channelId={channel.id}
+                    permissions={[isPrivate ? Permissions.MANAGE_PRIVATE_CHANNEL_PROPERTIES : Permissions.MANAGE_PUBLIC_CHANNEL_PROPERTIES]}
+                >
+                    {children}
+                </ChannelPermissionGate>
+            );
+        }
+    }
```

Comment:
On second iteration we can move the whole `ChannelPermissionGate` to the `SetHeaderButton` component (see `components/post_view/channel_intro_message/set_header_button.tsx`)

It can then look like this:

```suggestion
    const renderButtons = teamIsGroupConstrained && !(isReadOnly || channelIsArchived);
    const setHeaderButton = renderButtons ? <SetHeaderButton channel={channel}/> : null;
    const boardCreateButton = renderButtons ? <BoardsButton boardComponent={boardComponent}/> : null;
```

Reply:
done, with the caveat that teamIsGroupConstrained is not added here because it is used conditionally in `teamInviteLink`.  I'll check if that code can be refactored..